add RPC (-regtest only) for testing package policy

glozow opened the pull request:
It would be nice for LN/wallet/app devs to test out package policy, package RBF, etc., but the only interface to do so right now is through unit tests. This PR adds a `-regtest` only RPC interface so people can test by submitting raw transaction data. It is regtest-only, as it would be unsafe/confusing to create an actual mainnet interface while package relay doesn't exist.

Note that the functional tests are there to ensure the RPC interface is working properly; they aren't for testing policy itself. See src/test/txpackage_tests.cpp.

glozow committed:
[rpc] add new submitpackage RPC

It could be unsafe/confusing to create an actual mainnet interface while
package relay doesn't exist. However, a regtest-only interface allows
wallet/application devs to test current package policies.

glozow committed:
[functional test] submitrawpackage RPC

glozow commented:
TODOs (probably for future PRs):

- Automatically build the package using the child transaction, i.e. when we already have unconfirmed parents in the mempool or `mapWallet`. I expect to have this before package relay (users shouldn't be expected to construct packages themselves imo), but just haven't implemented it yet. LMK if this is something you want now.
- Add `maxfeerate` to allow users to set a maximum package feerate. This just requires adding a `TestPackage` codepath separate from testmempoolaccept so we can say "just kidding" if the validation succeeds.

luke-jr commented:
>It could be unsafe/confusing to create an actual mainnet interface while package relay doesn't exist.

How so?

> However, it's currently very inconvenient for wallet/application devs to try to test current package policies. 

They shouldn't need to. No assumptions should be made about user package policies...

MarcoFalke commented on src/rpc/mempool.cpp:
```diff
@@ -682,6 +810,7 @@ void RegisterMempoolRPCCommands(CRPCTable& t)
         {"blockchain", &getmempoolinfo},
         {"blockchain", &getrawmempool},
         {"blockchain", &savemempool},
+        {"hidden", &submitrawpackage,},
```
```suggestion
        {"hidden", &submitrawpackage},
```

nit?

glozow replied:
Done

MarcoFalke commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,133 @@ static RPCHelpMan savemempool()
+            HelpExampleCli("submitrawtransaction", "[rawtx1, rawtx2]")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
```
```suggestion
        {
```

nit: I accidentally didn't fix the whitespace when creating this file, but I think it is fine for new code to use the right amount of spaces.

glozow replied:
Done

maflcko reviewed (commented):
only nits

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,133 @@ static RPCHelpMan savemempool()
+                {
+                    {RPCResult::Type::OBJ, "xxxx", "transaction wtxid", {
+                        {RPCResult::Type::STR_HEX, "txid", "The transaction hash in hex"},
+                        {RPCResult::Type::STR_HEX, "wtxid", "The transaction witness hash in hex"},
```
If you're already using "wtxid" as a key, maybe not adding it as  "value" ?

glozow replied:
Because it's possible that the wtxid is different from the key, if there is a same-txid-different-witness tx in the mempool. I could make this more clear maybe?

glozow replied:
Ok I've renamed it to `other-wtxid` and only use it when communicating that a same-txid-different-witness tx was found. You're right that the wtxid key was redundant otherwise.

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,133 @@ static RPCHelpMan savemempool()
+        case PackageValidationResult::PCKG_POLICY:
+        case PackageValidationResult::PCKG_MEMPOOL_ERROR:
+        {
+            throw JSONRPCTransactionError(TransactionError::INVALID_PACKAGE,
```
Well sounds `PackageValidationResult::PCKG_MEMPOOL_ERROR` has its equivalent `TransactionError::MEMPOOL_ERROR` ? Or other reason to use `INVALID_PACKAGE` ?

glozow replied:
Good point, `TransactionError::MEMPOOL_ERROR` is more appropriate here. Switched.

glozow replied:
Good catch, swapped

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,133 @@ static RPCHelpMan savemempool()
+    return RPCHelpMan{"submitrawpackage",
+        "\nSubmit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only). Experimental, API may be unstable.\n"
+        "\nThe package will be validated and accepted to mempool if it passes consensus and mempool policy rules. Throws a JSONRPCError if anything goes wrong.\n"
+        "\nWarning: successful submission does not mean the transaction will propagate to other nodes on the network.\n",
```
To inform the package user, you could link back to `doc/policy/packages.md` somewhere.

glozow replied:
Added

ariard reviewed (commented):
Minors, otherwise SGTM.

ariard commented:
> They shouldn't need to. No assumptions should be made about user package policies...

Why no assumptions should be made about user package policies ? Could you layout the reasoning underpinning that statement ?

I believe to be of a contrary position. I think the default core policy should be designed to be compatible with miner incentives. In the present case, accepting relayed packages should increase your mining fee reward. If we assume that the majority of users are economically rational, they aim for their mempools contents to reflect the state of the next blocks and thus they should relay packages. Under this assumption concerning the base layer, I think an upper layer like LN can expect a policy like package relay to be widely supported.

glozow commented:
Addressed comments from @MarcoFalke and @ariard. Also renamed s/submitrawpackage/submitpackage because I don't think we'd ever submit non-raw packages.

luke-jr commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+static RPCHelpMan submitpackage()
+{
+    return RPCHelpMan{"submitpackage",
+        "\nSubmit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only).\n"
```
Seems like this PR should just be a modification to sendrawtransaction

glozow replied:
sendrawtransaction is not regtest-only

luke-jr replied:
Nor should this be, IMO. Certainly not long-term.

But regardless, the regtest check can be part of the existing RPC too. (I don't think it'd be the first case of such either IIRC)

ariard reviewed (approved):
Code Review ACK 04bb600

ariard commented:
> Users should ideally decide their own policies. This is especially true when consideration is being made for security purposes.

Users deciding their own policies doesn't mean they won't spontaneously converge towards few common policies per type of bitcoin applications/hosts.

> They create miner incentives, by deciding their own policies (blocks which don't match the node policy will relay slower).

If we define further miner incentives as optimizing block fee reward, i don't think node policies _creates_ them. Nodes policies (full-rbf, package relay) are just _means_ to achieve the incentives. I'm not sure it's a sane node policy to reject consensus-valid blocks.

> There is no economically rational reason for this. Miners have a reason to match node policies, but nodes do not really benefit from matching miners.

If users decide their own policies with a high degree of arbitrary of miners can expect to match the node policies ? At the contrary, node benefit from their policies matching the miners to improve their visibility of the future blockchain state and block space demand. I think restrictions, which can be motivated for security reasons or host perfomances or whatever, comes with the trade-off of a worsen view of the blockchain.

1440000bytes commented:
> They shouldn't need to. No assumptions should be made about user package policies.

@luke-jr Agree. Any application assuming mempool policies is vulnerable. Although this RPC could be helpful for testing in general.

maflcko reviewed (commented):
left some nits, since you requested a review (not sure if you wanted that kind of review)

michaelfolkson commented:
> Seems like this PR should just be a modification to sendrawtransaction IMO. (If there's a reason to make it regtest-only, that can still be done without a new RPC method)
> 
> But regardless, the regtest check can be part of the existing RPC too. (I don't think it'd be the first case of such either IIRC)

For the benefit of other reviewers this seems on the face of it to be a reasonable counter-suggestion to what this PR is doing that is worth exploring.

The rest of the conversation seems to be a rehashing of the "Lightning security ideally wouldn't rely on mempool policy" and "One of the motivations for having a mempool is having a set of already validated transactions that the node expects to be mined in an upcoming block"

jnewbery commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+{
+    return RPCHelpMan{"submitpackage",
+        "\nSubmit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only).\n"
+        "\nThe package will be validated and accepted to mempool if it passes consensus and mempool policy rules. Throws a JSONRPCError if anything goes wrong.\n"
```
"anything goes wrong" is  quite vague. Does it mean "if the transaction or package is rejected for consensus or policy reasons"?

glozow replied:
Hopefully now clearer

jnewbery commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+    return RPCHelpMan{"submitpackage",
+        "\nSubmit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only).\n"
+        "\nThe package will be validated and accepted to mempool if it passes consensus and mempool policy rules. Throws a JSONRPCError if anything goes wrong.\n"
+        "\nExperimental. API may be unstable. Refer to docs/policy/packages.md for documentation on package policies.\n"
```
```suggestion
        "\nThis RPC is experimental and the interface may be unstable. Refer to doc/policy/packages.md for documentation on package policies.\n"
```

glozow replied:
Thanks, taken

jnewbery commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+    NodeContext& node = EnsureAnyNodeContext(request.context);
+    CTxMemPool& mempool = EnsureMemPool(node);
+    CChainState& chainstate = EnsureChainman(node).ActiveChainstate();
+    const auto package_result = WITH_LOCK(::cs_main, return ProcessNewPackage(chainstate, mempool, txns, /* test_accept */ false));
```
```suggestion
    const auto package_result = WITH_LOCK(::cs_main, return ProcessNewPackage(chainstate, mempool, txns, /*test_accept=*/false));
```

glozow replied:
Done

jnewbery commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+                    throw JSONRPCTransactionError(TransactionError::MEMPOOL_REJECTED,
+                        strprintf("%s failed: %s", tx->GetHash().ToString(), it->second.m_state.GetRejectReason()));
+                }
+            }
```
Should we either `Assume(false)` here or raise a `JSONRPCTransactionError(TransactionError::MEMPOOL_ERROR)`?  There shouldn't be a `PackageValidationResult::PCKG_TX` without one of the transactions having an invalid status.

glozow replied:
Good point, added `Assume(false)` since it would be a code mistake

jnewbery commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+                        {RPCResult::Type::STR_HEX, "txid", "The transaction hash in hex"},
+                        {RPCResult::Type::STR_HEX, "other-wtxid", /*optional=*/ true, "The wtxid of a different transaction with the same txid but different witness found in the mempool. This means the submitted transaction was ignored."},
+                        {RPCResult::Type::NUM, "vsize", "Virtual transaction size as defined in BIP 141."},
+                        {RPCResult::Type::OBJ, "fees", "Transaction fees (only present if result=\"valid\")", {
```
```suggestion
                        {RPCResult::Type::OBJ, "fees", "Transaction fees", {
```

(result is always valid if this rpc returns)

glozow replied:
Ah, leftover from previous iteration, thanks for catching

jnewbery commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +306,80 @@ def test_rbf(self):
+
+        self.log.info("Submitrawpackage only allows packages of 1 child with its parents")
+        # Chain of 3 transactions has too many generations
+        (chain_hex, _) = create_raw_chain(node, self.coins.pop(), self.address, self.privkeys, 3)
```
No need for this to be a tuple:

```suggestion
        chain_hex, _ = create_raw_chain(node, self.coins.pop(), self.address, self.privkeys, 3)
```

glozow replied:
Done

jnewbery commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +306,80 @@ def test_rbf(self):
+        package_hex.append(child_hex)
+        package_txns.append(tx_from_hex(child_hex))
+
+        testres_package = node.testmempoolaccept(rawtxs=package_hex)
+        submitres_package = node.submitpackage(package=package_hex)
```
These names seems a bit confusing to me. What do you think about `testmempoolaccept_result` and `submitpackage_result`?

glozow replied:
Done

jnewbery commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +306,80 @@ def test_rbf(self):
+            # Grab this result from the submitres
+            submitres_tx = submitres_package["tx-results"][testres_tx["wtxid"]]
+            assert_equal(submitres_tx["txid"], testres_tx["txid"])
+            if "allowed" in testres_tx and testres_tx["allowed"]:
```
Is this always true?

glozow replied:
Not if we have partial submissions - added that test case and an explanation comment

jnewbery commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +306,80 @@ def test_rbf(self):
+
+        # With no deduplications, the package feerate is just the ancestor feerate of the child.
+        child_txid = testres_package[-1]["txid"]
+        if "package-feerate" in submitres_package:
```
Is this always true?

glozow replied:
Not if we have partial submissions - added that test case and an explanation comment

jnewbery reviewed (commented):
Concept and approach ACK. Just some minor style comments.

This could definitely have more extensive testing (eg submitting packages with entries already in the mempool, invalid package topologies, packages containing invalid transactions, etc), but it's fine to leave that for a follow-up PR. It looks like some of the test code has been written with that flexibility in mind, which I think can be removed if it's not actually being used yet.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+static RPCHelpMan submitpackage()
+{
+    return RPCHelpMan{"submitpackage",
+        "\nSubmit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only).\n"
```
Removed

glozow replied:
Wait I'm confused. Are there supposed to be newlines?

MarcoFalke replied:
Currently yes, newlines are needed within text, but not as a prefix for the description. (My bad, I've edited the suggested change above)

MarcoFalke replied:
We could also do auto-formatting of newlines, but maybe this can be done later

glozow replied:
Ok based on the new suggestion, I've added newlines at the end, but not at the beginning. Hopefully that's correct

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -669,6 +670,136 @@ static RPCHelpMan savemempool()
+            HelpExampleCli("submitrawtransaction", "[rawtx1, rawtx2]")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+    {
```
I assume you meant to indent all of it? Did that in the last push.

glozow commented:
Rebased to use `NONFATAL_UNREACHABLE`

glozow commented:
Thanks @MarcoFalke @jnewbery, took your comments

ariard reviewed (commented):
> For the benefit of other reviewers this seems on the face of it to be a reasonable counter-suggestion to what this PR is doing that is worth exploring.

FWIW, one argument can be made that make it easier to refine a new dedicated interface for future package issuer step by step as we experiment it. Of course, I think we could also extend `sendrawtransaction` with a regtest flag and package-only arguments. But that starts to be messy and we have more odds to interfere with `sendrawtransaction` usage, likely one of the most used RPC call across the Bitcoin ecosystem

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,138 @@ static RPCHelpMan savemempool()
+                        }},
+                    }}
+                }},
+                {RPCResult::Type::STR_AMOUNT, "package-feerate", /*optional=*/ true, "package feerate used for feerate checks in " + CURRENCY_UNIT + " per KvB. Excludes transactions which were deduplicated or accepted individually."},
```
Yeah I think that's a good idea for the future, see https://github.com/bitcoin/bitcoin/pull/24836#issuecomment-1096948837

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,138 @@ static RPCHelpMan savemempool()
+                UniValue::VARR,
+            });
+            const UniValue raw_transactions = request.params[0].get_array();
+            if (raw_transactions.size() < 1 || raw_transactions.size() > MAX_PACKAGE_COUNT) {
```
I think it's good to have a max number of transactions on the rpc. Could be 500 or something, but 25 is quite natural since it'll get rejected anyway if it's more than that. I could make a new number instead of `MAX_PACKAGE_COUNT` but again, more natural to just use a constant we already have.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,138 @@ static RPCHelpMan savemempool()
+        },
+        RPCExamples{
+            HelpExampleCli("testmempoolaccept", "[rawtx1, rawtx2]") +
+            HelpExampleCli("submitrawtransaction", "[rawtx1, rawtx2]")
```
Sorry yes, sendrawtransaction. Fixed

instagibbs commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,138 @@ static RPCHelpMan savemempool()
+        "This RPC is experimental and the interface may be unstable. Refer to doc/policy/packages.md for documentation on package policies."
+        "Warning: successful submission does not mean the transaction will propagate to other nodes on the network. Until package relay exists, they will each be relayed individually.",
+        {
+            {"package", RPCArg::Type::ARR, RPCArg::Optional::NO, "An array of raw transactions.",
```
mention the required topology?

instagibbs commented:
concept ACK

w0xlt reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/24836/commits/fbc6b8613eb4e6fd59fa45979aeff71e152c78e7

glozow commented:
Last push
- fixed the newline stuff in the description https://github.com/bitcoin/bitcoin/pull/24836#discussion_r899215072
- added a sentence about individual broadcast https://github.com/bitcoin/bitcoin/pull/24836#discussion_r899130155
- re-added `BroadcastTransaction()` (not sure if I lost that in a rebase or something)

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,138 @@ static RPCHelpMan savemempool()
+        "Submit a package of raw transactions (serialized, hex-encoded) to local node (-regtest only)."
+        "The package will be validated according to consensus and mempool policy rules. If all transactions pass, they will be accepted to mempool."
+        "This RPC is experimental and the interface may be unstable. Refer to doc/policy/packages.md for documentation on package policies."
+        "Warning: successful submission does not mean the transaction will propagate to other nodes on the network. Until package relay exists, they will each be relayed individually.",
```
Added

t-bast reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/24836/commits/980464989449b3b703c9a11e398b2b246dee163a

I modified [eclair](https://github.com/acinq/eclair) to use 0-fee commitment transactions and verified that I could use this RPC to submit packages containing a commitment transaction and a child anchor transaction. The RPC is easy to use and the data it returns is helpful.

> Seems like this PR should just be a modification to sendrawtransaction IMO.

I think this is a good point given the restrictions made to the topology of allowed packages. Because we're only considering packages of 1 child with multiple parents, we could use `sendrawtransaction` with the serialized child transaction and a new argument containing the package parents.

For now, a separate RPC that's regtest-only (and can thus be replaced / updated without caring too much about backwards-compatibility) is likely a better choice, but when the time comes to expose a mainnet RPC it could make sense to revisit that.

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,147 @@ static RPCHelpMan savemempool()
+            }
+            for (const auto& tx : txns) {
+                std::string err_string;
+                const auto err = BroadcastTransaction(node, tx, err_string, 0, true, true);
```
I think one way to exercise the `BroadcastTransaction` here could be to extend `rpc_packages.py` adding a `add_p2p_connection()` and waiting for an inv to be received with the package transactions txids.

glozow replied:
Added.

ariard reviewed (approved):
Code Review ACK 9804649

Changes since last review :
- `RPCHelpMan` updates to warn about the lack of package propagation
- `RPCResult` updates around "wtxid" and "other-wtxid"
- `HelpExampleCli` updates from "submitrawtransaction" to "sendrawtransaction"
- re-add the `BroadcastTransaction` (would be nice to test coverage as suggested)
- `rpc_package.py` comments and logs corrections

Edit: I checked my previous reviews, effectively `BroadcastTransaction()` wasn't present. From my understanding, without p2p packages, the odds of low-fee parent package propagating are low, so it didn't really matter if a `BroadcastTransaction()` call wasn't present or not. As this RPC is -regtest only and there is already a warning about expected lack of propagation, I think it doesn't matter if we introduce `BroadcastTransaction()` in this RPC now or latter.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,147 @@ static RPCHelpMan savemempool()
+                        }
+                    }
+                }
+                tx_result_map.pushKV(tx->GetWitnessHash().GetHex(), result_inner);
```
I think wtxids would be better in general. Txids for replaced txns is basically a legacy thing, it's just using what's returned in `MempoolAcceptResult` which is also used in testmempoolaccept. It would be impossible to have multiple replaced transactions with the same txids since they must have all been in mempool at the start. Perhaps one day we can change this for both RPCs?

Wtxids for indexing into the map because (1) it's more precise and (2) there actually is a possibility of same-txid-different-witness, i.e. see the "other-wtixd" result.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,147 @@ static RPCHelpMan savemempool()
+        },
+        RPCExamples{
+            HelpExampleCli("testmempoolaccept", "[rawtx1, rawtx2]") +
+            HelpExampleCli("sendrawtransaction", "[rawtx1, rawtx2]")
```
Thanks! Fixed.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -729,6 +730,147 @@ static RPCHelpMan savemempool()
+            for (const auto& tx : txns) {
+                std::string err_string;
+                const auto err = BroadcastTransaction(node, tx, err_string, 0, true, true);
+                if (err != TransactionError::OK) {
```
Note that we only do this after all the transactions have already been submitted successfully, so the error should be rare (in fact, kind of wondering if I should just put `NONFATAL_UNREACHABLE`). I've added the number of transactions broadcasted to the error string though.

instagibbs replied:
good enough for now

glozow commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +306,86 @@ def test_rbf(self):
+        self.assert_equal_package_results(node, testmempoolaccept_result, submitpackage_result)
+
+        # Package feerate is calculated for the remaining transactions after deduplication and
+        # individual submission.  If only 0 or 1 transaction is left, e.g. because all transactions
```
Added. Also note that we have more detailed tests for internal logic wrt package feerate in txpackage_tests.cpp

t-bast reviewed (approved):
Tested ACK against eclair https://github.com/bitcoin/bitcoin/pull/24836/commits/e866f0d0666664885d4c15c79bf59cc59975887a

ariard reviewed (approved):
Code Review ACK e866f0d0

Changes since last review:
- updates `sendrawtransaction` to `submitpackage` in `HelpExampleCli`
- yield a JSON rpc error with the number of transaction in package
- few functional test updates, notably adding a `test_submit_cpfp` case

instagibbs commented on test/functional/rpc_packages.py:
```diff
@@ -306,5 +309,127 @@ def test_rbf(self):
+        assert "package-feerate" in submitpackage_result
+        assert_fee_amount(DEFAULT_FEE, rich_parent_result["vsize"] + child_result["vsize"], submitpackage_result["package-feerate"])
+
+        # The node will broadcast each transaction, still abiding by its peer's fee filter
```
future work: would be nice if "abiding by its peer's fee filter" had test coverage rather than a comment

instagibbs reviewed (approved):
code review ACK e866f0d0666664885d4c15c79bf59cc59975887a

code changes as expected pert: https://github.com/bitcoin/bitcoin/pull/24836#issuecomment-1167147432
